src/Data/OffHeap.hs

```diff
@@ -0,0 +1,117 @@
+{-# LANGUAGE GADTs #-}
```

Comment:
I think `Data.OffHeap` is not a good name. I either doesn't mean anything to people, or is at best confusing because off-heap data is still on-heap, it's just not on on-GC-heap. We have an existing `Foreign.*` hierarchy, which among other things already provides `malloc`. This module is essentially a different API for `malloc`, so let's insert ourselves into the same hierarchy!

Call it e.g. `Foreign.Marshal.Alloc.Linear`. Now, you could argue that this is a different beast because it's pure. But I think this module should *not* be pure, making that point moot. At any rate, you cannot use the `Storable` type class if you're going to make this interface pure. `Storable` instances make no guarantees about referential transparency. Because methods are in `IO`, they can have arbitrary effects and that would still be fine. In fact `Storable` has no laws.

Also, I'm not sure you need the `Pool` concept anymore if you stick to `(R)IO`.